internal/controller/stages/control_flow_stages.go

```diff
@@ -0,0 +1,603 @@
+					predicate.GenerationChangedPredicate{},
+					kargo.RefreshRequested{},
+				),
+			),
+		).
+		Build(r)
+	if err != nil {
+		return fmt.Errorf("error building control flow Stage reconciler: %w", err)
+	}
+
+	// Configure the watches.
+	// Changes to these objects that match the constraints from the predicates
+	// will enqueue a reconciliation for the related Stage(s).
+
+	// Watch for Freight that are directly available from a Warehouse.
+	if err := c.Watch(
+		source.Kind(
+			mgr.GetCache(),
+			&kargoapi.Freight{},
+			&downstreamStageEnqueuer[*kargoapi.Freight]{
+				kargoClient: mgr.GetClient(),
+			},
+		),
+	); err != nil {
+		return fmt.Errorf("unable to watch Freight from upstream Stages: %w", err)
+	}
+
+	// Watch for Freight that have been verified in upstream Stages.
+	if err := c.Watch(
+		source.Kind(
+			mgr.GetCache(),
+			&kargoapi.Freight{},
+			&downstreamStageEnqueuer[*kargoapi.Freight]{
+				kargoClient:          mgr.GetClient(),
+				forControlFlowStages: true,
+			},
+		),
+	); err != nil {
+		return fmt.Errorf("unable to watch Freight verified in upstream Stages: %w", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the given control flow Stage.
+func (r *ControlFlowStageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"namespace", req.NamespacedName.Namespace,
+		"stage", req.NamespacedName.Name,
+		"controlFlow", true,
+	)
+	ctx = logging.ContextWithLogger(ctx, logger)
+
+	// Find the Stage.
+	stage := &kargoapi.Stage{}
+	if err := r.client.Get(ctx, req.NamespacedName, stage); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Safety check: do not reconcile Stages that are not control flow Stages.
+	if !stage.IsControlFlow() {
+		return ctrl.Result{}, nil
+	}
+
+	// Handle deletion of the Stage.
+	if !stage.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, r.handleDelete(ctx, stage)
+	}
+
+	// Ensure the Stage has a finalizer and requeue if it was added.
+	// The reason to requeue is to ensure that a possible deletion of the Stage
+	// directly after the finalizer was added is handled without delay.
+	if ok, err := kargoapi.EnsureFinalizer(ctx, r.client, stage); ok || err != nil {
+		return ctrl.Result{Requeue: ok}, err
+	}
+
+	// Reconcile the Stage.
+	logger.Debug("reconciling Stage")
+	newStatus, reconcileErr := r.reconcile(ctx, stage, time.Now())
+	logger.Debug("done reconciling Stage")
+
+	// Patch the status of the Stage.
+	if err := kubeclient.PatchStatus(ctx, r.client, stage, func(status *kargoapi.StageStatus) {
+		*status = newStatus
+	}); err != nil {
+		// Prioritize the reconcile error if it exists.
+		if reconcileErr != nil {
+			logger.Error(err, "failed to update Stage status after reconciliation error")
+			return ctrl.Result{}, reconcileErr
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to update Stage status: %w", err)
+	}
+	return ctrl.Result{}, reconcileErr
```

Comment:
I need to test this more thoroughly, but I believe there is no actual reason for us to requeue on an interval here because we should be informed by a watcher or change event.

Reply:
Tested this, and what's expected from it is triggered just fine without it.